File: cpp/cpp_100-999/875.cc
// C++
// leetcode 875
// https://leetcode.cn/problems/koko-eating-bananas/

class Solution {
public:
	int get_h(vector<int>& piles, int speed) {
		int res = 0;
		for(auto &x : piles) {
			res += (x / speed) + (x % speed == 0 ? 0 : 1);
		}
		return res;
	}

    int minEatingSpeed(vector<int>& piles, int h) {
    	int max_bananas = 0;
    	for(auto &x : piles)
    		max_bananas = max(max_bananas, x);
    	int l = 1, r = max_bananas;
    	while(l <= r) {
    		// l always ok
    		// cout<<l<<", "<<r<<endl;
    		int mid = (l + r + 1) / 2;
    		if(get_h(piles, mid) <= h) {
    			r = mid - 1;
    		} else {
    			l = mid + 1;
    		}
    	}
    	return l;
    }
};

File: Go/Go_100-999/875.go
// Go
// leetcode 875
// https://leetcode.cn/problems/koko-eating-bananas/

func minEatingSpeed(piles []int, h int) int {
	m := 0
	for _, v := range piles {
		if v > m {
			m = v
		}
	}
	l, r := 1, m
	for l <= r {
		mid := (l + r + 1) >> 1
		if get_h(&piles, mid) <= h {
			r = mid - 1
		} else {
			l = mid + 1
		}
	}
	return l
}

func get_h(p *[]int, k int) int {
	res := 0
	for _, v := range *p {
		res += v / k
		if v%k != 0 {
			res++
		}
	}
	return res
}